proposals/handler.go

```diff
@@ -0,0 +1,380 @@
+		logger.With().Warning("failed to process ballot", log.Err(err))
+		return err
+	}
+
+	if err := h.checkTransactions(ctx, &p); err != nil {
+		logger.With().Warning("failed to fetch proposal TXs", log.Err(err))
+		return err
+	}
+
+	h.logger.WithContext(ctx).With().Debug("proposal is syntactically valid")
+	reportProposalMetrics(&p)
+
+	// TODO: add Proposal/Ballot/TXs to mesh
+
+	if err := h.mesh.AddProposal(&p); err != nil {
+		logger.With().Error("failed to save proposal", log.Err(err))
+		return fmt.Errorf("save proposal: %w", err)
+	}
+
+	return nil
+}
+
+func (h *Handler) processBallot(ctx context.Context, b *types.Ballot, logger log.Log) error {
+	if h.mesh.HasBallot(b.ID()) {
+		logger.Info("known ballot")
+		return nil
+	}
+	logger.With().Info("new ballot", b.Fields()...)
+
+	if err := h.checkBallotSyntacticValidity(ctx, b); err != nil {
+		logger.With().Error("ballot syntactically invalid", log.Err(err))
+		return fmt.Errorf("syntactic-check ballot: %w", err)
+	}
+
+	reportBallotMetrics(b)
+	return nil
+}
+
+func (h *Handler) checkBallotSyntacticValidity(ctx context.Context, b *types.Ballot) error {
+	h.logger.WithContext(ctx).With().Debug("checking proposal syntactic validity")
+
+	if err := h.checkBallotDataIntegrity(b); err != nil {
+		h.logger.WithContext(ctx).With().Error("ballot integrity check failed", log.Err(err))
+		return err
+	}
+
+	if err := h.checkBallotDataAvailability(ctx, b); err != nil {
+		h.logger.WithContext(ctx).With().Error("ballot data availability check failed", log.Err(err))
+		return err
+	}
+
+	if eligible, err := h.validator.CheckEligibility(ctx, b); err != nil || !eligible {
+		h.logger.WithContext(ctx).With().Error("ballot eligibility check failed", log.Err(err))
+		return errNotEligible
+	}
+
+	h.logger.WithContext(ctx).With().Debug("ballot is syntactically valid")
+	return nil
+}
+
+func (h *Handler) checkBallotDataIntegrity(b *types.Ballot) error {
+	if b.AtxID == *types.EmptyATXID || b.AtxID == h.cfg.GoldenATXID {
+		return errInvalidATXID
+	}
+
+	if b.BaseBallot == types.EmptyBallotID {
+		return errMissingBaseBallot
+	}
+
+	if b.RefBallot == types.EmptyBallotID {
+		// this is the smesher's first Ballot in this epoch, should contain EpochData
+		if b.EpochData == nil {
+			return errMissingEpochData
+		}
+		if len(b.EpochData.Beacon) == 0 {
+			return errMissingBeacon
+		}
+		if len(b.EpochData.ActiveSet) == 0 {
+			return errEmptyActiveSet
+		}
+		// check for duplicate ATXIDs in active set
+		set := make(map[types.ATXID]struct{}, len(b.EpochData.ActiveSet))
+		for _, atx := range b.EpochData.ActiveSet {
+			if _, exist := set[atx]; exist {
+				return errDuplicateATX
+			}
+			set[atx] = struct{}{}
+		}
+	} else {
+		if b.EpochData != nil {
+			return errUnexpectedEpochData
+		}
+	}
+
+	if err := h.checkExceptions(b, h.cfg.MaxExceptions); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *Handler) checkExceptions(ballot *types.Ballot, max int) error {
```

Comment:
it doesn't have to be a method, it deliberately passes all dependencies using arguments, so that there is no implicit stuff going on

Reply:
done